Replace the ROCm specific multiple NCCL manager assert with a one-sho…

rsanthanam-amd opened the pull request:
…t warning log message.

deven-amd commented on tensorflow/core/nccl/nccl_manager.cc:
```diff
@@ -178,9 +178,13 @@ struct NcclManager::Collective : public core::RefCounted {
-          "ROCm cannot use multi-node NCCL collectives on a single node");
+      static absl::once_flag once;
+      absl::call_once(once, [] {
+        LOG(WARNING) << "ROCm cannot use multi-node NCCL collectives on a single node";
```
we need to change the wording of the warning message..something like

```
More than one concurrent instance of NcclManager has been instantiated on the same node, this can sometimes result in a race condition leading to a system hang.
```

@jeffdaily what say you?

deven-amd reviewed (changes requested):
In the commit message, add full details of the error scenario that is prompting this change, including instructions on how to reproduce it and the analysis we have done so far. 

It is very likely that we are going to revisit this issue again in the future, and having all the details captured will come in handy.

Alternately, you can append all the details to this issue in our internal repo - https://github.com/ROCmSoftwarePlatform/frameworks-internal/issues/81 , and merely add a link to it in the commit message

rsanthanam-amd committed:
This change replaces the original assert for detecting multiple
NCCL managers in favor of a warning log message.

The original assert was added to protect against a potential deadlock
scenario (that is specific to ROCm) when multiple hosts are emulated
using a single host.  This can happen in this scenario because two
different NCCL streams can map to the same hardware queue and block
each other from making forward progress.

This deadlock scenario seems to only occur in this specific scenario.

However, the assert was preventing other scenarios involving multiple
NCCL managers from executing even though they did not cause a deadlock.

For example, TF creates a single eager context (which creates a NCCL manager)
prior to running a model in eager mode.  However, other eager contexts can be
created for single use purposes which results in additional NCCL managers being
created but not used duing the model run.  In these scenarios, deadlock is
impossible and thus the creation of multiple NCCL managers should be allowed
to proceed.

The specific case which is prompting this change is that in addition to the
regular eager context, an additional context is created by the MLIR constant
folding optimization to compute the constant at compile time so that it can
be used as a replacement later on the compilation process.  This additional
context is only used once and cannot cause a deadlock.

This change enables these cases to execute successfully.

In the future, the original assert will have to be reinstated with
additional logic restricting its execution to the specific scenario(s)
which cause the deadlock.

rsanthanam-amd commented:
Upgraded the commit message and warning log message.
